src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -25,121 +25,113 @@ private sealed class ChunkedEncodingReadStream : HttpContentReadStream
 
-            private async Task<bool> TryGetNextChunkAsync()
+            public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken)
             {
-                Debug.Assert(_chunkBytesRemaining == 0);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    // Cancellation requested.
+                    return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
+                }
 
-                // Read the start of the chunk line.
-                _connection._allowedReadLineBytes = MaxChunkBytesAllowed;
-                ArraySegment<byte> line = await _connection.ReadNextLineAsync().ConfigureAwait(false);
+                if (_connection == null || destination.Length == 0)
+                {
+                    // Response body fully consumed or the caller didn't ask for any data.
+                    return new ValueTask<int>(0);
+                }
 
-                // Parse the hex value.
-                if (!Utf8Parser.TryParse(line.AsReadOnlySpan(), out ulong chunkSize, out int bytesConsumed, 'X'))
+                // Try to consume from data we already have in the buffer.
+                int bytesRead = ReadChunksFromConnectionBuffer(destination.Span);
+                if (bytesRead > 0)
                 {
-                    throw new IOException(SR.net_http_invalid_response);
+                    return new ValueTask<int>(bytesRead);
                 }
-                else if (bytesConsumed != line.Count)
+
+                // Nothing available to consume.  Fall back to I/O.
+                return ReadAsyncCore(destination, cancellationToken);
+            }
+
+            private async ValueTask<int> ReadAsyncCore(Memory<byte> destination, CancellationToken cancellationToken)
+            {
+                // Should only be called if ReadChunksFromConnectionBuffer returned 0.
+
+                Debug.Assert(_connection != null);
+                Debug.Assert(destination.Length > 0);
+
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
+                try
                 {
-                    // There was data after the chunk size, presumably a "chunk extension".
-                    // Allow tabs and spaces and then stop validating once we get to an extension.
-                    int offset = line.Offset + bytesConsumed, end = line.Count - bytesConsumed + line.Offset;
-                    for (int i = offset; i < end; i++)
+                    while (true)
                     {
-                        char c = (char)line.Array[i];
-                        if (c == ';')
+                        if (_state == ParsingState.Ending)
                         {
-                            break;
+                            await ConsumeTrailingHeaders().ConfigureAwait(false);
+                            Finish();
```

Comment:
Nit: can we move the Finish call to ConsumeTrailingHeaders?

Reply:
Sure